bftengine/src/bftengine/SigManager.cpp

```diff
@@ -271,18 +251,103 @@ bool SigManager::verifySig(
+  ConcordAssert(replicasInfo_.isIdOfReplica(myId_) || replicasInfo_.isRoReplica());
+  concord::crypto::ISigner* rawSigner = nullptr;
+  std::shared_ptr<EdDSAMultisigSigner> signer;
+  if (ReplicaConfig::instance().singleSignatureScheme) {
+    signer = CryptoManager::instance().getSigner(seq);
+    rawSigner = signer.get();
+  } else {
+    rawSigner = mySigner_.get();
+  }
+  auto result = rawSigner->signBuffer(data, dataLength, outSig);
+  LOG_DEBUG(GL, "Signing as replica with " << KVLOG(myId_, seq, rawSigner->signatureLength(), result));
+  return result;
 }
 
-size_t SigManager::sign(const char* data, size_t dataLength, char* outSig) const {
-  return sign(reinterpret_cast<const uint8_t*>(data), dataLength, reinterpret_cast<uint8_t*>(outSig));
+size_t SigManager::sign(SeqNum seq, const char* data, size_t dataLength, char* outSig) const {
+  return sign(seq, reinterpret_cast<const uint8_t*>(data), dataLength, reinterpret_cast<uint8_t*>(outSig));
+}
+
+bool SigManager::verifyReplicaSig(PrincipalId replicaID,
+                                  const concord::Byte* data,
+                                  size_t dataLength,
+                                  const concord::Byte* sig,
+                                  uint16_t sigLength) const {
+  LOG_DEBUG(GL, KVLOG(myId_));
+  ConcordAssert(replicasInfo_.isIdOfReplica(myId_) || replicasInfo_.isRoReplica());
+  ConcordAssert(replicasInfo_.isIdOfReplica(replicaID));
+  int i = 0;
+  for (auto [seq, multisigVerifier] : CryptoManager::instance().getLatestVerifiers()) {
+    UNUSED(seq);
+    if (multisigVerifier.get() == nullptr) {
+      continue;
+    }
+    auto& verifier = multisigVerifier->getVerifier(replicaID);
+    LOG_DEBUG(GL,
+              "Validating as replica with: " << KVLOG(
+                  myId_, replicaID, i, verifier.getPubKey(), sigLength, verifier.signatureLength()));
+    if (verifier.verifyBuffer(data, dataLength, sig, sigLength)) {
+      LOG_DEBUG(GL, "Validation Successful " << KVLOG(myId_, replicaID, i, sigLength, verifier.signatureLength()));
+      return true;
+    } else {
+      LOG_DEBUG(
+          GL,
+          "Validation failed as replica with: " << KVLOG(myId_, replicaID, i, sigLength, verifier.signatureLength()));
+    }
+    i++;
+  }
+  LOG_WARN(GL, "Validation failed with all cryptosystems" << KVLOG(myId_, replicaID, i));
+  return false;
+}
+
+// verify using the two last keys, once the last key's checkpoint is reached, the previous key is removed
+bool SigManager::verifySig(
+    PrincipalId pid, const concord::Byte* data, size_t dataLength, const concord::Byte* sig, uint16_t sigLength) const {
+  if (replicasInfo_.isIdOfReplica(pid) && ReplicaConfig::instance().singleSignatureScheme) {
+    return verifyReplicaSig(pid, data, dataLength, sig, sigLength);
+  }
+
+  return verifyNonReplicaSig(pid, data, dataLength, sig, sigLength);
+}
+
+bool SigManager::verifyOwnSignature(const concord::Byte* data,
```

Comment:
why don't we validate the signature with a validator?

Reply:
IDK, I just extracted a logic which was used in the preprocessor.